File: DriftGame/uracer-desktop/src/com/bitfire/uracer/utils/Timer.java
package com.bitfire.uracer.utils;

import com.badlogic.gdx.utils.TimeUtils;

public final class Timer {

    private static final float oneOnOneBillion = 1.0f / 1000000000.0f;
    private long nsStartTime;
    private boolean stopped;
    private long nsStopTime;
    private float elapsedSecs;

    public Timer() {
        reset();
    }

    public void pause() {
        stopped = true;
        nsStopTime = TimeUtils.nanoTime();
    }

    public void resume() {
        stopped = false;
    }

    public void reset() {
        stopped = false;

        // abs
        nsStartTime = TimeUtils.nanoTime();
        nsStopTime = 0;
    }

    public void update() {
        long now = (stopped ? nsStopTime : TimeUtils.nanoTime());
        elapsedSecs = (float) (now - nsStartTime) * oneOnOneBillion;
    }

    /**
     * Returns the elapsed time in seconds.
     */
    public float elapsed() {
        return elapsedSecs;
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/utils/Timer.kt
package com.bitfire.uracer.utils

import com.badlogic.gdx.utils.TimeUtils

class Timer {

    private var nsStartTime = 0L
    private var stopped = false
    private var nsStopTime = 0L
    private var elapsedSecs = 0F

    init {
        reset()
    }

    fun pause() {
        stopped = true
        nsStopTime = TimeUtils.nanoTime()
    }

    fun resume() {
        stopped = false
    }

    fun reset() {
        stopped = false
        nsStartTime = TimeUtils.nanoTime()
        nsStopTime = 0
    }

    fun update() {
        val now = (if (stopped) nsStopTime else TimeUtils.nanoTime())
        elapsedSecs = (now - nsStartTime).toFloat() * oneOnOneBillion
    }

    /**
     * Returns the elapsed time in seconds.
     */
    fun elapsed() = elapsedSecs

    companion object {
        private const val oneOnOneBillion = 1.0F / 1000000000.0F
    }
}
